File: javascript/src/js/sorting/quickSort.js
function pivot(arr, i, j) {
    if(i === j) {
        return j;
    }
    var pivot = i;
    i = i+1;
    while( i <= j ) {
      
        while(arr[i] <= arr[pivot]) {
            i++;
        }
        while(arr[j] > arr[pivot]) {
            j--;
        }

        if( i < j ) {
            var temp = arr[i];
            arr[i] = arr[j];
            arr[j] = temp;
            i++;
            j--;
       }
      
    }
    console.log("Value of j => ", j, arr[j]);
    var temp = arr[pivot];
    arr[pivot] = arr[j];
    arr[j] = temp;
    console.log(arr);
    return j
}

function quickSort(array,left,right) {
     if( left >= right ) {
         console.log( "Left and Right => " + left , right);
         return array;
     }

     
     var position = pivot(array,left,right)
     quickSort(array,left,position-1);
     quickSort(array,position+1,right);
     return array;
}

// let arr = [3,44,38,5,47,15,36,26,27,2,46,4,19,50,4];
// let arr = [8,5,2,9,5,6,3]
// let arr = [1]
// let arr = [1, 2];
// let arr = [2, 1];
// let arr = [1, 3, 2];
// let arr = [3, 1, 2];
// let arr = [1, 2, 3];
let arr = [-4,5,10,8,-10,-6,-4,-2,-5,3,5,-4,-5,-1,1,6,-7,-6,-7,8]
// let arr = [-7, 2, 3, 8, -10, 4, -6, -10, -2, -7, 10, 5, 2, 9, -9, -5, 3, 8]
// let arr = [5, -2, 2, -8, 3, -10, -6, -1, 2, -2, 9, 1, 1];
// let arr = [2,-2,-6,-10,10,4,-8,-1,-8,-4,7,-4,0,9,-9,0,-9,-9,8,1,-4,4,8,5,1,5,0,0,2,-10]

// let arr = [ 8, 5, 2, 9, 5, 6, 3 ]
let result = quickSort(arr,0,arr.length-1);
console.log(result);

File: java/src/main/java/com/google/QuickSort.java
package com.google;

import java.io.*;
import java.util.*;

/*
 * To execute Java, please define "static void main" on a class
 * named Solution.
 *
 * If you need more classes, simply define them inline.
 */

class QuickSort {
    public static void main(String[] args) {

        // int [] arr = new int[] { 32,12,31,5,22,1};
        int [] arr = new int[] { 22,1,32,3,45,4,41};
        // int [] arr = new int[] { 1,2,3,4,5,6,7};
        // int [] arr = new int[] { 7,6,5,4,3,2,1};
        // int [] arr = new int[] { 1,12,4,23,32};

        helper(arr);

    }

    public static void helper(int [] arr) {
        int left = 0;
        int right = arr.length-1;
        quickSort(arr,left,right);
        System.out.println("Completed Quick sort " + Arrays.toString(arr));
    }

    public static void quickSort(int [] arr, int left , int right) {
        if(left > right) {
            return;
        }
        int mid = partition(arr,left,right);
        System.out.println("Mid element => " + mid);
        quickSort(arr,left,mid-1);
        quickSort(arr,mid+1,right);
        // System.out.println(Arrays.toString(arr));

    }

    public static int partition(int [] arr, int left, int right) {

        int pIndex = left;
        int pivot = arr[right];

        for(int i = pIndex; i < right; i++) {
            if(arr[i] <= pivot) {
                swap(arr,pIndex,i);
                pIndex++;
            }

        }

        swap(arr,right,pIndex);
        System.out.println(Arrays.toString(arr));
        return pIndex;

    }

    public static void swap(int [] arr, int i, int j) {
        // System.out.println("i  : :: j " + i + " : " + j );
        System.out.println("Swapped elements ==>  " + Arrays.toString(arr));
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }

}
